linkerd/app/inbound/src/direct.rs

```diff
@@ -114,18 +126,40 @@ where
         .push(metrics.transport.layer_accept())
         .push_map_target(TcpAccept::from)
         .push(tls::NewDetectTls::layer(
-            // TODO set ALPN on these connections to indicate support
-            // for the transport header.
-            local_identity,
+            local_identity.map(WithTransportHeaderAlpn),
             config.detect_protocol_timeout,
         ))
         .check_new_service::<listen::Addrs, I>()
         .into_inner()
 }
 
-// === impl NonOpaqueRefused ===
+// === impl WithTransportHeaderAlpn ===
+
+impl Into<tls::server::Config> for &'_ WithTransportHeaderAlpn {
+    fn into(self) -> tls::server::Config {
+        // Copy the underlying TLS config and set an ALPN value.
+        //
+        // TODO: Avoid cloning the server config for every connection. It would
+        // be preferable if rustls::ServerConfig wrapped individual fields in an
+        // Arc so they could be overridden independently.
```

Comment:
does this need an upstream issue?

Reply:
https://github.com/ctz/rustls/issues/431 was closed the other day, so there may not be interest in doing this. We can probably improve the situation internally but it's too big of a change to worry bout for now. Per-connection cloning in special cases isn't great but it won't kill us.

Reply:
:/ yeah, i don't think it's a *huge* deal. it might be possible to make the issue a little clearer to the rustls folks?